### PR TITLE
Blog post for producer tuning

### PR DESCRIPTION
**Pull request by PaulRMellor**
Signed-off-by: prmellor <pmellor@redhat.com>

New post to describe the configuration properties to use for producer optimization

### CODE REVIEW

**Comment by PaulRMellor** [comment]
Included a couple of changes related to community feedback

File: _includes/documentation/documentation-version-band.html [hunk 1]
```diff
@@ -26,9 +26,9 @@ <h3>Documentation for Version {{site.data.releases.operator[0].version}} - lates
       </p>
       <p>
         Strimzi Kafka operators Master<br>
-        <a href="{{site.baseurl}}/docs/operators/master/overview.html">Overview guide</a> | 
-        <a href="{{site.baseurl}}/docs/operators/master/quickstart.html">Quick Start guide</a> | 
-        <a href="{{site.baseurl}}/docs/operators/master/deploying.html">Deploying Strimzi</a> |
+        <a href="{{site.baseurl}}/docs/operators/master/overview.html">Overview guide</a> |
+        <a href="{{site.baseurl}}/docs/operators/master/quickstart.html">Quick Start guide</a> |
+        <a href="{{site.baseurl}}/docs/operators/master/deploying.html">Deploying and Upgrading Strimzi</a> |
```
Comment (scholzj) [inline _includes/documentation/documentation-version-band.html:31]: This seems to break the line. The book name is also used on more places. Maybe we should leave it to another PR where we can play a bit more with the design?

**Review by scholzj — COMMENTED** [review]
I think the content is solid. But you have written it as a many small paragraphs - often with just one sentence. I think that looks for me a bit weird and is not completely easy to follow the flow (have a look at the preview: https://deploy-preview-197--strimzi.netlify.app/blog/2020/10/02/producer-tuning/) ...was that intentional?

File: _posts/2020-10-02-producer-tuning.md [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+
+Think long enough about this, and you might find competing requirements. For example, by maximizing throughput you might also increase latency.
+
+Be prepared to make adjustments to your adjustments.
+
+### How is your producer performing?
+
+It's only when you have been monitoring the performance of your producers for some time that you can gauge how best to tune their performance.
+
+To begin with, you might start with a basic producer configuration in development as a benchmark. When you start to analyze producer metrics to see how the producers actually perform in typical production scenarios, you can make incremental changes and make comparisons until you hit the sweet spot.
```
Comment (ppatierno) [inline _posts/2020-10-02-producer-tuning.md:34]: maybe mentioning the producer metrics we can put a reference to the producer monitoring documentation on the Kafka website? https://kafka.apache.org/documentation/#producer_sender_monitoring

File: _posts/2020-10-02-producer-tuning.md [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+
+```properties
+bootstrap.servers=localhost:9092
+key.serializer=org.apache.kafka.common.serialization.StringSerializer
+value.serializer=org.apache.kafka.common.serialization.StringSerializer
+client.id=my-client
+compression.type=gzip
+```
+
+This configuration specifies the bootstrap address for connection to the Kafka cluster, and the serializers that transform the key and value of a message.
```
Comment (ppatierno) [inline _posts/2020-10-02-producer-tuning.md:55]: I would be clear saying that the serializers transform the key and value from String to the corresponding raw bytes representation to be sent on the wire. Saying just "transform" doesn't sound clear to what IMHO.

File: _posts/2020-10-02-producer-tuning.md [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+spec:
+  partitions: 1
+  replicas: 3
+  config:
+    min.insync.replicas: 2
+    #...
+```
+Using a topic replication factor of 3, and 2 in-sync replicas on other brokers, the producer can continue unaffected if a single broker is unavailable.
+
+If a second broker becomes unavailable, the producer won’t receive acknowledgments and won’t be able to produce more messages.
```
Comment (ppatierno) [inline _posts/2020-10-02-producer-tuning.md:126]: It is true if the producer still sends with acks=all but it can still send if using the others type of acknowledgement.

**Comment by PaulRMellor** [comment]
> I think the content is solid. But you have written it as a many small paragraphs - often with just one sentence. I think that looks for me a bit weird and is not completely easy to follow the flow (have a look at the preview: https://deploy-preview-197--strimzi.netlify.app/blog/2020/10/02/producer-tuning/) ...was that intentional?

Always thinking of ease of reading, but it's a bit too broken up. Will brings some of the paragraphs together.

File: _includes/documentation/documentation-version-band.html (hunk 1 above)
Comment (PaulRMellor) [reply _includes/documentation/documentation-version-band.html:31]: Reverted

File: _posts/2020-10-02-producer-tuning.md (hunk 2 above)
Comment (PaulRMellor) [reply _posts/2020-10-02-producer-tuning.md:34]: good idea. Added:

> If you want to read more about performance metrics for monitoring Kafka producers, see Kafka's [Producer Sender Metrics](https://kafka.apache.org/documentation/#producer_sender_monitoring).

File: _posts/2020-10-02-producer-tuning.md (hunk 3 above)
Comment (PaulRMellor) [reply _posts/2020-10-02-producer-tuning.md:55]: :+1: 

This configuration specifies the bootstrap address for connection to the Kafka cluster, and the serializers that transform the key and value of a message from from a String to its corresponding raw byte data representation.

File: _posts/2020-10-02-producer-tuning.md (hunk 4 above)
Comment (PaulRMellor) [reply _posts/2020-10-02-producer-tuning.md:126]: :+1: 

> If a second broker becomes unavailable, using `acks=all` the producer won’t receive acknowledgments and won’t be able to produce more messages.

**Review by scholzj — APPROVED** [review]

**Review by ppatierno — APPROVED** [review]

File: _posts/2020-10-02-producer-tuning.md [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+Here we see an example configuration showing idempotence enabled, and used with `max.in.flight.requests.per.connection` and `acks=all`.
+
+```properties
+# ...
+enable.idempotence=true
+max.in.flight.requests.per.connection=5
+acks=all
+retries=2147483647
+```
+The `retries` property sets the number of retries when resending a failed message request. That impressive-looking number is the default and maximum value.
```
Comment (tombentley) [inline _posts/2020-10-02-producer-tuning.md:142]:
```suggestion
The `retries` property sets the number of retries when resending a failed message request. While that number might look impressive, what we're saying is, in effect, "retry forever".
```

File: _posts/2020-10-02-producer-tuning.md [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+
+> **Delivery timeout**
+>
+>If you are using `delivery.timeout.ms` in your producer configuration, producer requests will fail before the number of retries has been used if the timeout expires before a successful acknowledgment. The `delivery.timeout.ms` sets a limit on the time to wait for an acknowledgment of the success or failure to deliver a message. You can choose to leave `retries` unset and use `delivery.timeout.ms` to perform a similar function instead.
+
+There is a performance cost to introducing additional checks to the order of delivery.
+
+If you prefer not to use `acks=all` and idempotency,
+a lighter option is to set the number of in-flight requests to 1 (the default is 5) to preserve ordering.
```
Comment (tombentley) [inline _posts/2020-10-02-producer-tuning.md:151]: I wouldn't use "lighter", as the lightness really depends on the requirements. Not pipelining comes with its own costs.

File: _posts/2020-10-02-producer-tuning.md [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+transactional.id=_UNIQUE-ID_
+transaction.timeout.ms=900000
+# ...
+```
+
+The transactional ID is registered with the Kafka cluster on the first operation, as well as a producer `epoch` checkpoint number, which is used to identify the active producer instance.
+
+Why might we want to identify the active producer? Say an application determines that a producer has failed and creates a new producer instance to restart a transaction. If both producers are now sending messages, duplicate records are being created and we have lost our exactly once integrity.
+
+By specifying a transaction ID, if a new producer instance starts, older instances of the producer are identified by their older `epoch` number and _fenced-off_ by Kafka so that their messages are no included. This maintains the integrity of the message passing by ensuring that there is only ever one valid producer with the transactional ID. Each `transactional.id` should be used for a unique set of topic partitions.
```
Comment (tombentley) [inline _posts/2020-10-02-producer-tuning.md:186]:
```suggestion
By specifying a transaction ID, if a new producer instance starts, older instances of the producer are identified by their older `epoch` number and _fenced-off_ by Kafka so that their messages are not included. This maintains the integrity of the message passing by ensuring that there is only ever one valid producer with the transactional ID. Each `transactional.id` should be used for a unique set of topic partitions.
```

File: _posts/2020-10-02-producer-tuning.md [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+
+You can map topic partition names to transactional IDs, or compute the transactional ID from the topic partition names using a function that avoids collisions.
+
+### Optimizing throughput and latency
+
+Depending on your objective, Kafka offers a number of configuration parameters and techniques for tuning producer performance for throughput and latency.
+
+Usually, the requirement of a system is to satisfy a particular throughput target for a proportion of messages within a given latency.
+For example, targeting 500,000 messages per second with 95% of messages being acknowledged within 2 seconds.
```
Comment (tombentley) [inline _posts/2020-10-02-producer-tuning.md:195]:
```suggestion
For example, targeting 50,000 messages per second with 95% of messages being acknowledged within 2 seconds.
```

File: _posts/2020-10-02-producer-tuning.md [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+If you think compression is worthwhile, the best type of compression to use will depend on the messages being sent.
+
+> **Adding threads**
+>
+> Compression is handled on the thread calling `KafkaProducer.send()`, so if the latency of this method matters for your application you can add more more threads.
+
+#### Pipelining messages
+
+Pipelining might sound like it has something to do with surfing the famous _Pipeline_ reef in Hawaii,
+but it's actually the equally impressive method of sending more requests from producers before the response to a previous request has been received.
```
Comment (tombentley) [inline _posts/2020-10-02-producer-tuning.md:275]:
```suggestion
but it's actually the sending more requests from producers before the response to a previous request has been received.
```

File: _posts/2020-10-02-producer-tuning.md [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+### Adapt to survive
+
+Fine tuning your producers helps alleviate performance issues.
+But don't be tempted to make a few adjustments and think your work is done.
+
+You should consider fine tuning as part of a continual optimization process. Monitor Kafka regularly. Look for changing trends in usage and investigate how fine tuning can help your Kafka deployment adapt. But, as you'll know, this is only one half of the story.
+
+Next time we'll look at how you can optimize your consumers.
+
+Be sure to check back for the next installment.
```
Comment (tombentley) [inline _posts/2020-10-02-producer-tuning.md:307]: Should we promise this?

File: _posts/2020-10-02-producer-tuning.md [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+```properties
+# ...
+delivery.timeout.ms=120000
+partitioner.class=my-custom-partitioner
+# ...
+```
+
+### Adapt to survive
+
+Fine tuning your producers helps alleviate performance issues.
```
Comment (tombentley) [inline _posts/2020-10-02-producer-tuning.md:300]:
```suggestion
Fine-tuning your producers helps alleviate performance issues.
```

File: _posts/2020-10-02-producer-tuning.md [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+But don't be tempted to make a few adjustments and think your work is done.
+
+You should consider fine tuning as part of a continual optimization process. Monitor Kafka regularly. Look for changing trends in usage and investigate how fine tuning can help your Kafka deployment adapt. But, as you'll know, this is only one half of the story.
```
Comment (tombentley) [inline _posts/2020-10-02-producer-tuning.md:303]:
```suggestion
You should consider fine-tuning as part of a continual optimization process. Monitor Kafka regularly. Look for changing trends in usage and investigate how fine-tuning can help your Kafka deployment adapt. But, as you'll know, this is only one half of the story.
```

File: _posts/2020-10-02-producer-tuning.md [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+spec:
+  partitions: 1
+  replicas: 3
+  config:
+    min.insync.replicas: 2
+    #...
+```
+Using a topic replication factor of 3, and 2 in-sync replicas on other brokers, the producer can continue unaffected if a single broker is unavailable.
+
+If a second broker becomes unavailable, using `acks=all` the producer won’t receive acknowledgments and won’t be able to produce more messages.
```
[the remark this answers is not in the document]
Comment (PaulRMellor) [reply _posts/2020-10-02-producer-tuning.md:120]: I think we'll need to stick with US/global english for acknowledgement

File: _posts/2020-10-02-producer-tuning.md (hunk 5 above)
Comment (PaulRMellor) [reply _posts/2020-10-02-producer-tuning.md:142]: Thanks for clarifying. Useful.

File: _posts/2020-10-02-producer-tuning.md (hunk 6 above)
Comment (PaulRMellor) [reply _posts/2020-10-02-producer-tuning.md:151]: changed to _...another option is..._

File: _posts/2020-10-02-producer-tuning.md (hunk 10 above)
Comment (PaulRMellor) [reply _posts/2020-10-02-producer-tuning.md:307]: I think so. I have a _to do_ to expand the section in the guide to include consumers. Might need some help though.

**Review by tombentley — APPROVED** [review]
